src/vmm/src/devices/virtio/vhost_user.rs

```diff
@@ -257,3 +464,517 @@ impl VhostUserHandle {
         Ok(())
     }
 }
+
+#[cfg(test)]
+mod tests {
+    #![allow(clippy::undocumented_unsafe_blocks)]
+    #![allow(clippy::cast_possible_truncation)]
+
+    use utils::tempfile::TempFile;
+
+    use super::*;
+    use crate::vstate::memory::{FileOffset, GuestAddress, GuestMemoryExtension};
+
+    #[test]
+    fn test_new() {
+        struct MockMaster {
+            sock: UnixStream,
+            max_queue_num: u64,
+            is_owner: std::cell::UnsafeCell<bool>,
+        }
+
+        impl VhostUserHandleBackend for MockMaster {
+            fn from_stream(sock: UnixStream, max_queue_num: u64) -> Self {
+                Self {
+                    sock,
+                    max_queue_num,
+                    is_owner: std::cell::UnsafeCell::new(false),
+                }
+            }
+
+            fn set_owner(&self) -> Result<(), vhost::Error> {
+                unsafe { *self.is_owner.get() = true };
+                Ok(())
+            }
+        }
+
+        let max_queue_num = 69;
+
+        let tmp_dir = utils::tempdir::TempDir::new().unwrap();
+        let tmp_dir_path_str = tmp_dir.as_path().to_str().unwrap();
+        let tmp_socket_path = format!("{tmp_dir_path_str}/tmp_socket");
+
+        unsafe {
+            let socketfd = libc::socket(libc::AF_UNIX, libc::SOCK_STREAM, 0);
```

Comment:
do we need to create a real socket in the unit test?

Reply:
but that's `VhostUserHandleImpl<Master>` that connects, not `VhostUserHandleImpl<MockMaster>`, and we're testing the latter, aren't we?

Reply:
`VhostUserHandleImpl<MockMaster>` also connects to the socket (creates a `UnixStream`). The difference is that `MockMaster` does nothing with that connection.
```rust
pub fn new(socket_path: &str, num_queues: u64) -> Result<Self, VhostUserError> {
        let stream = UnixStream::connect(socket_path).map_err(VhostUserError::Connect)?;

        let vu = T::from_stream(stream, num_queues);
        vu.set_owner().map_err(VhostUserError::VhostUserSetOwner)?;

        Ok(Self {
            vu,
            socket_path: socket_path.to_string(),
        })
}
```

Reply:
yeah, I see. It doesn't look very nice, but probably isn't a big issue.